diff --git a/CMakeLists.txt b/CMakeLists.txt
index 134523e7..52fec96e 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -189,7 +189,7 @@ if(MSVC)
 else(MSVC)
   string(APPEND CMAKE_CXX_FLAGS " -Wall")
   string(APPEND CMAKE_CXX_FLAGS " -Wextra")
-  string(APPEND CMAKE_CXX_FLAGS " -Werror")
+  #string(APPEND CMAKE_CXX_FLAGS " -Werror")
   string(APPEND CMAKE_CXX_FLAGS " -Wno-deprecated-declarations")
   string(APPEND CMAKE_CXX_FLAGS " -Wimplicit-fallthrough")
   target_compile_options(fbgemm_avx2 PRIVATE
@@ -208,6 +208,11 @@ else(MSVC)
     src/FbgemmI8DepthwiseAvx2.cc
     src/UtilsAvx2.cc
     PROPERTIES COMPILE_FLAGS "-Wno-uninitialized")
+  set_source_files_properties(
+    src/RefImplementations.cc
+    src/UtilsAvx2.cc
+    src/UtilsAvx512.cc
+    PROPERTIES COMPILE_FLAGS "-Wno-uninitialized -O0 -g0")
   set_source_files_properties(src/PackMatrix.cc
     PROPERTIES COMPILE_FLAGS "-Wno-infinite-recursion")
   # Workaround for https://gcc.gnu.org/bugzilla/show_bug.cgi?id=80947
