diff --git a/cmake/microkernels.cmake b/cmake/microkernels.cmake
index 93c7a1127..c845e6ad5 100644
--- a/cmake/microkernels.cmake
+++ b/cmake/microkernels.cmake
@@ -1157,6 +1157,13 @@ SET(ALL_AVX2_MICROKERNEL_SRCS
 
 SET(ALL_AVX512AMX_MICROKERNEL_SRCS)
 
+set_source_files_properties(
+  src/x16-packw/gen/x16-packw-x8-gemm-goi-avx2-u16-prfm.c
+  src/x16-packw/gen/x16-packw-x8-gemm-goi-avx2-u16.c
+  src/x16-packw/gen/x16-packw-x16-gemm-goi-avx2-u16-prfm.c
+  src/x16-packw/gen/x16-packw-x16-gemm-goi-avx2-u16.c
+  PROPERTIES COMPILE_FLAGS "-O0 -g0")
+
 SET(ALL_AVX512F_MICROKERNEL_SRCS
   src/f32-dwconv/gen/f32-dwconv-3p16c-minmax-avx512f-acc2.c
   src/f32-dwconv/gen/f32-dwconv-3p16c-minmax-avx512f.c
